### ISSUE TITLE
Rate limited on the OpenAI embeddings API, sleeping before retrying...

### ISSUE THREAD

**Issue by BitcoinCode** [Azure-Samples/azure-search-openai-demo#431]
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is for a: (mark with an `x`)
```
- [x ] bug report -> please search issues before submitting
- [ ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### Minimal steps to reproduce
> prepdocs.py '..\data\*' --storageaccount myaccount --container mycontainer --searchservice mysearch --index myindex -v

### Any log messages given by the failure
>Ensuring search index gptkbindex exists
Search index gptkbindex already exists
Processing files...
Processing '../pdf/xxx.pdf'
        Uploading blob for page 0 -> xxx-0.pdf
Extracting text from '../pdfs/xxx.pdf' using Azure Form Recognizer
Indexing sections from 'xxx.pdf' into search index 'gptkbindex'
Splitting '../pdfs/xxx.pdf' into sections
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Traceback (most recent call last):
  File "/path/.local/lib/python3.10/site-packages/tenacity/__init__.py", line 382, in __call__
    result = fn(*args, **kwargs)
  File "/path/scripts/prepdocs.py", line 216, in compute_embedding
    return openai.Embedding.create(engine=args.openaideployment, input=text)["data"][0]["embedding"]
  File "/path/.local/lib/python3.10/site-packages/openai/api_resources/embedding.py", line 33, in create
    response = super().create(*args, **kwargs)
  File "/path/.local/lib/python3.10/site-packages/openai/api_resources/abstract/engine_api_resource.py", line 149, in create
    ) = cls.__prepare_create_request(
  File "/path/.local/lib/python3.10/site-packages/openai/api_resources/abstract/engine_api_resource.py", line 83, in __prepare_create_request
    raise error.InvalidRequestError(
openai.error.InvalidRequestError: Must provide an 'engine' or 'deployment_id' parameter to create a <class 'openai.api_resources.embedding.Embedding'>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/path/scripts/prepdocs.py", line 367, in <module>
    index_sections(os.path.basename(filename), sections)
  File "/path/scripts/prepdocs.py", line 262, in index_sections
    for s in sections:
  File "/path/scripts/prepdocs.py", line 208, in create_sections
    section["embedding"] = compute_embedding(content)
  File "/path/.local/lib/python3.10/site-packages/tenacity/__init__.py", line 289, in wrapped_f
    return self(f, *args, **kw)
  File "/path/.local/lib/python3.10/site-packages/tenacity/__init__.py", line 379, in __call__
    do = self.iter(retry_state=retry_state)
  File "/path/.local/lib/python3.10/site-packages/tenacity/__init__.py", line 326, in iter
    raise retry_exc from fut.exception()
tenacity.RetryError: RetryError[<Future at 0x7f4635e70ac0 state=finished raised InvalidRequestError>]

### Expected/desired behavior
> To create vector values so I can use the hybrid drop-down to chat to my docs 

### OS and Version?
>Linux (Ubuntu 22.04)

### azd version?
> run `azd version` and copy paste here.

### Versions
>azd version 1.1.0 (commit ea9cb12575734ee6a5f99c4d415c1a51d6f32d3e)

### Mention any other details that might be useful

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

**Comment by shulkx** [comment]
I met the same problem.

**Comment by ayoubazaouyat** [comment]
@shulkx  can you give an example I didn't understand you properly? should I add them to my `azd up` command ?

**Comment by ikka86** [comment]
The same problem is occurring.
Where can I get--openaideployment?
Which files need to be changed?

**Comment by shulkx** [comment]
I didn't use the `azd up` because it contains everything inside with less customizable.
I run the [prepdocs.py](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/scripts/prepdocs.py) manually by passing in parameters to specific services (e.g., blob storage, form recognizer key, etc.,). Therefore, I can choose to pass in the above 3 parameters I mentioned.
After the indexing process is done, directly run `azd deploy` to deploy the webapp.

**Comment by pamelafox** [comment]
The recent change to vector search also changed the prepdocs scripts, like prepdocs.sh:
https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/scripts/prepdocs.sh

The two added arguments were:

```
--openaiservice "$AZURE_OPENAI_SERVICE" --openaideployment "$AZURE_OPENAI_EMB_DEPLOYMENT"
```

If you run prepdocs.sh or prepdocs.ps1, then they will get those values automatically from your current azd environment.

If you are running prepdocs.py manually, then you'll need to find those values in .azure/CURRENT-ENV-NAME/.env and pass them in.

**Comment by olsgaard** [comment]
@pamelafox 

I'm running the latest version of Azure-search-openai-demo (initialized from template yesterday), so I have those changes, but I still get the rate limit problem that results in an exception being raised as described by bitcoincode.

I have 147 documents containing only ~20k word or just under 120k characters, so I shouldn't even be hitting the rate limit.

**Comment by pamelafox** [comment]
Thanks for the details. We're looking into various approaches to work around the rate limiting. For now I'd suggest breaking the work up (like into multiple folders and running it on each folder). Sorry about the errors!

**Comment by ananthuraj-cresen** [comment]
Are there any updates regarding this issue?
ERROR: failed running post hooks: 'postprovision' hook failed with exit code: '1', Path: './scripts/prepdocs.sh'. : exit code: 1
"statusCode": 401, "message": "Unauthorized. Access token is missing, invalid, audience is incorrect

**Comment by GitTxb2** [comment]
Not sure if this is related to this issue thread:
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Rate limited on the OpenAI embeddings API, sleeping before retrying...
Traceback (most recent call last):
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\.venv\Lib\site-packages\openai\api_requestor.py", line 403, in handle_error_response
    error_data = resp["error"]
                 ~~~~^^^^^^^^^
KeyError: 'error'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\.venv\Lib\site-packages\tenacity\__init__.py", line 382, in __call__
    result = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\prepdocs.py", line 229, in compute_embedding
    return openai.Embedding.create(engine=args.openaideployment, input=text)["data"][0]["embedding"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\.venv\Lib\site-packages\openai\api_resources\embedding.py", line 33, in create
    response = super().create(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\.venv\Lib\site-packages\openai\api_resources\abstract\engine_api_resource.py", line 153, in create     
    response, _, api_key = requestor.request(
                           ^^^^^^^^^^^^^^^^^^
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\.venv\Lib\site-packages\openai\api_requestor.py", line 298, in request
    resp, got_stream = self._interpret_response(result, stream)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\.venv\Lib\site-packages\openai\api_requestor.py", line 700, in _interpret_response
    self._interpret_response_line(
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\.venv\Lib\site-packages\openai\api_requestor.py", line 763, in _interpret_response_line
    raise self.handle_error_response(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\.venv\Lib\site-packages\openai\api_requestor.py", line 405, in handle_error_response
    raise error.APIError(
openai.error.APIError: Invalid response object from API: '{ "statusCode": 500, "message": "Internal server error", "activityId": "699ff4be-563c-4f43-821a-5fa50ac1996d" }' (HTTP response code was 500)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\prepdocs.py", line 380, in <module>
    index_sections(os.path.basename(filename), sections)
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\prepdocs.py", line 275, in index_sections
    for s in sections:
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\prepdocs.py", line 221, in create_sections
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\.venv\Lib\site-packages\tenacity\__init__.py", line 289, in wrapped_f
    return self(f, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\.venv\Lib\site-packages\tenacity\__init__.py", line 379, in __call__
    do = self.iter(retry_state=retry_state)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\.venv\Lib\site-packages\tenacity\__init__.py", line 326, in iter
    raise retry_exc from fut.exception()
tenacity.RetryError: RetryError[<Future at 0x1f4c47d20d0 state=finished raised APIError>]

**Comment by GitTxb2** [comment]
> Not sure if this is related to this issue thread: Rate limited on the OpenAI embeddings API, sleeping before retrying... Rate limited on the OpenAI embeddings API, sleeping before retrying... Rate limited on the OpenAI embeddings API, sleeping before retrying... Rate limited on the OpenAI embeddings API, sleeping before retrying... Rate limited on the OpenAI embeddings API, sleeping before retrying... Rate limited on the OpenAI embeddings API, sleeping before retrying... Rate limited on the OpenAI embeddings API, sleeping before retrying... Rate limited on the OpenAI embeddings API, sleeping before retrying... Rate limited on the OpenAI embeddings API, sleeping before retrying... Rate limited on the OpenAI embeddings API, sleeping before retrying... Traceback (most recent call last): File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts.venv\Lib\site-packages\openai\api_requestor.py", line 403, in handle_error_response error_data = resp["error"] ~~~~^^^^^^^^^ KeyError: 'error'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last): File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts.venv\Lib\site-packages\tenacity__init__.py", line 382, in **call** result = fn(*args, **kwargs) ^^^^^^^^^^^^^^^^^^^ File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\prepdocs.py", line 229, in compute_embedding return openai.Embedding.create(engine=args.openaideployment, input=text)["data"][0]["embedding"] ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts.venv\Lib\site-packages\openai\api_resources\embedding.py", line 33, in create response = super().create(*args, **kwargs) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts.venv\Lib\site-packages\openai\api_resources\abstract\engine_api_resource.py", line 153, in create response, _, api_key = requestor.request( ^^^^^^^^^^^^^^^^^^ File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts.venv\Lib\site-packages\openai\api_requestor.py", line 298, in request resp, got_stream = self._interpret_response(result, stream) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts.venv\Lib\site-packages\openai\api_requestor.py", line 700, in _interpret_response self._interpret_response_line( File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts.venv\Lib\site-packages\openai\api_requestor.py", line 763, in _interpret_response_line raise self.handle_error_response( ^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts.venv\Lib\site-packages\openai\api_requestor.py", line 405, in handle_error_response raise error.APIError( openai.error.APIError: Invalid response object from API: '{ "statusCode": 500, "message": "Internal server error", "activityId": "699ff4be-563c-4f43-821a-5fa50ac1996d" }' (HTTP response code was 500)
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last): File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\prepdocs.py", line 380, in index_sections(os.path.basename(filename), sections) File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\prepdocs.py", line 275, in index_sections for s in sections: File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts\prepdocs.py", line 221, in create_sections ^^^^^^^^^^^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^^^^^^^^^^^ File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts.venv\Lib\site-packages\tenacity__init__.py", line 289, in wrapped_f return self(f, *args, **kw) ^^^^^^^^^^^^^^^^^^^^ File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts.venv\Lib\site-packages\tenacity__init__.py", line 379, in **call** do = self.iter(retry_state=retry_state) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "C:\Users\tyler\AZUREOPENI\azure-search-openai-demo\scripts.venv\Lib\site-packages\tenacity__init__.py", line 326, in iter raise retry_exc from fut.exception() tenacity.RetryError: RetryError[<Future at 0x1f4c47d20d0 state=finished raised APIError>]


And in the GUI I receive the following:

<img width="405" alt="Screenshot 2023-08-08 105046" src="https://github.com/Azure-Samples/azure-search-openai-demo/assets/119471993/4a71c253-3c42-42af-92f7-c6bbcdd7de18">

**Comment by vrajroutu** [comment]
@GitTxb2 

Azure OAI Outage in EAST-US effecting emedding models

FYI

`Your support case @####$#%$^#&%&$&&$  is related to an ongoing outage in your region`.

`STATUS:
In-Progress 8/8/2023 4:43:29 AM UTC`

`SUMMARY OF IMPACT:
You may experience sporadic issues accessing your resources. This could cause you to experience a 5xx error when calling Davinci model or Ada embedding model in EastUS region.`

`NEXT STEPS:
Azure OpenAI feature team are urgently investigating the root cause and coming up with a plan to resolve as soon as possible. We'll send an update once we know more.`

**Comment by GitTxb2** [comment]
@vrajroutu Thanks so much for letting me know. I was tearing my hair out! Thanks again

**Comment by ikka86** [comment]
@vrajroutu 

I am getting the same error as well.
Where can I find that information?
It is not listed at
https://azure.status.microsoft/ja-jp/status

**Comment by GitTxb2** [comment]
@ikka86 My CSP reports the issue has been fixed. I'll do some further testing today. 
![image](https://github.com/Azure-Samples/azure-search-openai-demo/assets/119471993/ad50812b-824f-43b9-ad7e-efbb26b5b852)

**Comment by pamelafox** [comment]
Closing this issue as it's been reported fixed in multiple venues. Sorry about that!

**Comment by pamelafox** [comment]
Sorry, re-opening since this issue was originally about the prepdocs.py issue, which folks may still run into.

**Comment by sunny5055** [comment]
> 

Hi Any update to this issue ? or any workaround at the moment ? 
I get the error even when adding a single document now .

**Comment by pamelafox** [comment]
There's a workaround in this PR: https://github.com/Azure-Samples/azure-search-openai-demo/pull/529

**Comment by BitcoinCode** [comment]
1- Clear the data from the data folder when doing azd up
2- Put the date on separate folder on your system 
3- Once the deployment done put the files back to your data folder and use prepdoc.ps1 not prepdoc.py as .sh file will have the right parameters.

for more info look at @pamelafox  comment in here https://github.com/Azure-Samples/azure-search-openai-demo/issues/431#issuecomment-1643661657

**Comment by smeury** [comment]
I also experienced this error this week, but just when I use it with the option --openaikey.

My workaround is to comment or delete "refresh_openai_token()" on line 234 in prepdocs.py

This refresh_openai_token() throws an error, which then leads to the false errormessage about the rate limited.

**Comment by pamelafox** [comment]
@smeury Ah that's actually a bit of a bug, I've proposed a fix in #573
